port/win/port_win.h

```diff
@@ -0,0 +1,576 @@
+//  Copyright (c) 2013, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under the BSD-style license found in the
+//  LICENSE file in the root directory of this source tree. An additional grant
+//  of patent rights can be found in the PATENTS file in the same directory.
+//
+// Copyright (c) 2011 The LevelDB Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file. See the AUTHORS file for names of contributors.
+//
+// See port_example.h for documentation for the following types/functions.
+
+#ifndef STORAGE_LEVELDB_PORT_PORT_WIN_H_
+#define STORAGE_LEVELDB_PORT_PORT_WIN_H_
+
+// Always want minimum headers
+#ifndef WIN32_LEAN_AND_MEAN
+#  define WIN32_LEAN_AND_MEAN
+#endif
+
+// Assume that for everywhere
+#undef PLATFORM_IS_LITTLE_ENDIAN
+#define PLATFORM_IS_LITTLE_ENDIAN true
+
+#include <windows.h>
+#include <string>
+#include <string.h>
+#include <mutex>
+#include <condition_variable>
+
+#include <stdint.h>
+
+#include "rocksdb/options.h"
+
+#ifndef strcasecmp
+#define strcasecmp _stricmp
+#endif
+
+// defined in stdio.h
+#ifndef snprintf
+#define snprintf _snprintf
+#endif
+
+typedef SSIZE_T ssize_t;
+
+// size_t printf formatting named in the manner of C99 standard formatting strings such as PRIu64
+// in fact, we could use that one
+#define ROCKSDB_PRIszt "Iu"
+
+#define __attribute__(A)
+
+#ifdef ZLIB
+#include <zlib.h>
+#endif
+
+#ifdef BZIP2
+#include <bzlib.h>
+#endif
+
+#if defined(LZ4)
+#include <lz4.h>
+#include <lz4hc.h>
+#endif
+
+#ifdef SNAPPY
+#include <snappy.h>
+#endif
+
+// Thread local storage on Linux
+// There is thread_local in C++11
+#define __thread __declspec(thread)
+
+#ifndef PLATFORM_IS_LITTLE_ENDIAN
+#define PLATFORM_IS_LITTLE_ENDIAN (__BYTE_ORDER == __LITTLE_ENDIAN)
+#endif
+
+namespace rocksdb {
+
+#define PREFETCH(addr, rw, locality)
+
+namespace port 
+{
+
+// For use at db/file_indexer.h kLevelMaxIndex
+const int LevelMaxIndex = INT32_MAX;
+
+const bool kLittleEndian = true;
+
+class CondVar;
+
+class Mutex 
+{
+public:
+    /* implicit */ Mutex(bool adaptive = false);
+    ~Mutex();
+
+    void Lock();
+    void Unlock();
+  
+    // this will assert if the mutex is not locked
+    // it does NOT verify that mutex is held by a calling thread
+    void AssertHeld();
+
+    std::unique_lock<std::mutex>& getLock() {
+        return lock;
+    }
+
+private:
+    friend class CondVar;
+    std::mutex m_mutex;
```

Comment:
I would hold for this for now. There is a possibility that Natham wants to implement our own mutex to improve write queue sychronization.